### PR TITLE
fix downloading of NPMs

### PR DESCRIPTION
**Pull request by brastak**
For #49 (part two, downloading fix)

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#55](https://github.com/artipie/npm-adapter/issues/55) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/bb6a7f44-ce86-4936-95e8-9e20ceb03667); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by victornoel** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/npm-adapter/issues/55#issuecomment-628214899))

@victornoel The user @victornoel[/z](https://www.0crat.com/u/victornoel) resigned from [#55](https://github.com/artipie/npm-adapter/issues/55), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/npm-adapter/issues/55#issuecomment-628214899))

@victornoel Job refused in 0 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
This pull request [#55](https://github.com/artipie/npm-adapter/issues/55) is assigned to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/f1aa132b-ac0f-4c7a-854c-91691bb0e528); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/main/java/com/artipie/npm/http/DownloadPackageSlice.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import org.reactivestreams.Publisher;
+
+/**
+ * Download package endpoint. Return package metadata, all tarball links will be rewritten
+ * based on requested URL.
+ *
+ * @since 0.6
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/npm/http/DownloadPackageSlice.java:62]: @brastak are you sure this suppression is required?

File: src/test/java/com/artipie/npm/NpmCommandsTest.java [hunk 2]
```diff
@@ -78,24 +85,31 @@ void tearDown() {
      * Test {@code npm publish} command works properly.
      * @throws IOException if fails
      * @throws InterruptedException if fails
+     * @throws ExecutionException if fails
      */
     @Test
     void npmPublishWorks() throws IOException, InterruptedException, ExecutionException {
```
Comment (olenagerasimova) [inline src/test/java/com/artipie/npm/NpmCommandsTest.java:91]: @brastak why this test is disabled on windows?

File: src/main/java/com/artipie/npm/http/RoutingAwareSlice.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.regex.Pattern;
+import org.reactivestreams.Publisher;
+
+/**
+ * Slice handles routing paths. It removes predefined routing path and passes the rest part
+ * to the underlying slice.
+ *
+ * @since 0.6
+ */
+public final class RoutingAwareSlice implements Slice {
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/npm/http/RoutingAwareSlice.java:40]: @brastak this class needs unit test, please, add todo

File: src/main/java/com/artipie/npm/http/DownloadPackageSlice.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        );
+    }
+
+    /**
+     * Read content and replace tarball links with absolute paths based on prefix.
+     * @param original Original JSON metadata content
+     * @param prefix Links prefix
+     * @return Transformed JSON metadata content
+     */
+    private static CompletionStage<? extends Content> processTarballs(
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/npm/http/DownloadPackageSlice.java:134]: @brastak static methods of this class can be extracted into separate class and properly unit-tested: for example, we could call new class `Tarbolls`, add `Content` as field and create method `update` with `prefix` parameter that will return `JsonObject`. I do not insist on refactoring it now, but please add todo for it.

File: src/main/java/com/artipie/npm/http/DownloadPackageSlice.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    // @checkstyle ReturnCountCheck (50 lines)
+    @Override
+    public Response response(final String line,
+        final Iterable<Map.Entry<String, String>> headers,
+        final Publisher<ByteBuffer> body) {
+        final String pkg = new PackageNameFromUrl(this.path, line).value();
+        final String host = StreamSupport.stream(headers.spliterator(), false)
+            .filter(e -> e.getKey().equalsIgnoreCase("Host"))
+            .findAny().orElseThrow(
+                () -> new RuntimeException("Could not find Host header in request")
```
[the remark this answers is not in the document]
Comment (brastak) [reply src/main/java/com/artipie/npm/http/DownloadPackageSlice.java:95]: @olenagerasimova Why? From javadoc: Signals that a method has been invoked at an illegal or inappropriate time.

File: src/test/java/com/artipie/npm/NpmCommandsTest.java (hunk 2 above)
Comment (brastak) [reply src/test/java/com/artipie/npm/NpmCommandsTest.java:91]: @olenagerasimova because windows fails on run npm publish command

**Comment by brastak** [comment]
@olenagerasimova ok, I will split this PR into 3. But it doesn't make development process faster in my opinion.

**Comment by olenagerasimova** [comment]
@brastak I do not think it makes any sense  now, I've already reviewed this one, you can add todo/answers and we'll merge it

**Comment by 0crat** [comment]
Job was finished in 10 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Pull request [#55](https://github.com/artipie/npm-adapter/issues/55) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Comment by sereshqua** [comment]
@brastak please make sure that all problems found by the code reviewer (@olenagerasimova) were addressed by the pull request author (YOU), which means that they were either fixed as the reviewer wanted or the reviewer was convinced that they should not be fixed

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Comment by g4s8** [comment]
@rultor release, tag=`0.6`

**Pull request closed without merging**